exonum/src/encoding/fields.rs

```diff
@@ -304,43 +304,82 @@ impl<'a> Field<'a> for DateTime<Utc> {
     }
 }
 
+const HEADER_SIZE: usize = 1;
+const BODY_SIZE: usize = mem::size_of::<Ipv6Addr>();
+const PORT_SIZE: usize = 2;
+
+const IPV4_HEADER: u8 = 0;
+const IPV6_HEADER: u8 = 1;
+
 // TODO add socketaddr check, for now with only ipv4
 // all possible (>6 bytes long) sequences is a valid addr (ECR-156).
 impl<'a> Field<'a> for SocketAddr {
     fn field_size() -> Offset {
         // FIXME: reserve space for future compatibility (ECR-156)
-        6
+        (HEADER_SIZE + BODY_SIZE + PORT_SIZE) as Offset
     }
 
     unsafe fn read(buffer: &'a [u8], from: Offset, to: Offset) -> Self {
-        let mut octets = [0u8; 4];
-        octets.copy_from_slice(&buffer[from as usize..from as usize + 4]);
-        let ip = Ipv4Addr::from(octets);
-        let port = LittleEndian::read_u16(&buffer[from as usize + 4..to as usize]);
-        SocketAddr::V4(SocketAddrV4::new(ip, port))
+        let addr_start = from as usize + HEADER_SIZE;
+        let ip = match buffer[from as usize] {
+            IPV4_HEADER => {
+                const ADDR_SIZE: usize = mem::size_of::<Ipv4Addr>();
+                let mut octets: [u8; ADDR_SIZE] = mem::uninitialized();
+                octets.copy_from_slice(&buffer[addr_start..addr_start + ADDR_SIZE]);
+                IpAddr::V4(Ipv4Addr::from(octets))
+            }
+            IPV6_HEADER => {
+                const ADDR_SIZE: usize = mem::size_of::<Ipv6Addr>();
+                let mut octets: [u8; ADDR_SIZE] = mem::uninitialized();
+                octets.copy_from_slice(&buffer[addr_start..addr_start + ADDR_SIZE]);
+                IpAddr::V6(Ipv6Addr::from(octets))
+            }
+            header => panic!("Unknown header `{:X}` for SocketAddr", header),
```

Comment:
I wonder if we should mark that as `unreachable`. In theory, we always need to `check` before `read`.

Reply:
The motivation behind it was that it is still theoretically possible to reach that `match` branch. By panicking there instead of `unreachable!()` we don't get any extra overhead (besides binary size I guess), so I'd rather leave it there just in case.